src/java/org/apache/cassandra/config/DurationSpec.java

```diff
@@ -244,153 +170,442 @@ public long to(TimeUnit targetUnit)
+         *
+         * @return this duration in number of minutes or {@code Integer.MAX_VALUE} if the number of minutes is too large.
+         */
+        public int toMinutes()
+        {
+            return Ints.saturatedCast(unit().toMinutes(quantity()));
+        }
     }
 
     /**
-     * Returns this duration in number of milliseconds as an {@code int}
-     *
-     * @return this duration in number of milliseconds or {@code Integer.MAX_VALUE} if the number of milliseconds is too large.
+     * Represents a duration used for Cassandra configuration. The bound is [0, Integer.MAX_VALUE) in seconds.
+     * If the user sets a different unit - we still validate that converted to seconds the quantity will not exceed
+     * that upper bound. (CASSANDRA-17571)
      */
-    public int toMillisecondsAsInt()
+    public final static class IntSecondsBound extends DurationSpec
     {
-        return Ints.saturatedCast(toMilliseconds());
-    }
-
-    @Override
-    public int hashCode()
-    {
-        // Milliseconds seems to be a reasonable tradeoff
-        return Objects.hash(unit.toMillis(quantity));
-    }
+        private static final Pattern VALUES_PATTERN = Pattern.compile(("\\d+"));
+
+        /**
+         * Creates a {@code DurationSpec.IntSecondsBound} of the specified amount. The bound is [0, Integer.MAX_VALUE) in seconds.
+         * The bound is [0, Integer.MAX_VALUE) in seconds.
+         *
+         * @param value the duration
+         */
+        public IntSecondsBound(String value)
+        {
+            super(value, SECONDS, Integer.MAX_VALUE);
+        }
 
-    @Override
-    public boolean equals(Object obj)
-    {
-        if (this == obj)
-            return true;
+        /**
+         * Creates a {@code DurationSpec.IntSecondsBound} of the specified amount in the specified unit.
+         * The bound is [0, Integer.MAX_VALUE) in seconds.
+         *
+         * @param quantity where quantity shouldn't be bigger than Integer.MAX_VALUE - 1 in seconds
+         * @param unit in which the provided quantity is
+         */
+        public IntSecondsBound(long quantity, TimeUnit unit)
+        {
+            super(quantity, unit, SECONDS, Integer.MAX_VALUE);
+        }
 
-        if (!(obj instanceof DurationSpec))
-            return false;
+        /**
+         * Creates a {@code DurationSpec.IntSecondsBound} of the specified amount in seconds.
+         * The bound is [0, Integer.MAX_VALUE) in seconds.
+         *
+         * @param seconds where seconds shouldn't be bigger than Integer.MAX_VALUE-1
+         */
+        public IntSecondsBound(long seconds)
+        {
+            this(seconds, SECONDS);
+        }
 
-        DurationSpec other = (DurationSpec) obj;
-        if (unit == other.unit)
-            return quantity == other.quantity;
+        /**
+         * Creates a {@code DurationSpec.IntSecondsBound} of the specified amount in seconds.
```

Comment:
Nit: Actually, this method accepts both a number of seconds without unit, or a number of whatevers with unit. I'd change the JavaDoc to reflect this, for example:
```suggestion
         * Creates a {@code DurationSpec.IntSecondsBound} of the specified amount in seconds, expressed either as the
         * number of seconds without unit, or as a regular quantity with unit.
```
I have tried to think on a more descriptive name for the method but nothing came out, I guess the JavaDoc should be enough.

Reply:
I like the idea, makes it more explicit, thanks!